alibi_detect/od/tests/test__knn/test__knn.py

```diff
@@ -0,0 +1,250 @@
+import pytest
+import numpy as np
+import torch
+
+from alibi_detect.od._knn import KNN
+from alibi_detect.od.pytorch.ensemble import AverageAggregator, TopKAggregator, MaxAggregator, \
+    MinAggregator, ShiftAndScaleNormalizer, PValNormalizer
+from alibi_detect.exceptions import NotFittedError, ThresholdNotInferredError
+
+from sklearn.datasets import make_moons
+
+
+def make_knn_detector(k=5, aggregator=None, normalizer=None):
+    knn_detector = KNN(
+        k=k, aggregator=aggregator,
+        normalizer=normalizer
+    )
+    x_ref = np.random.randn(100, 2)
+    knn_detector.fit(x_ref)
+    knn_detector.infer_threshold(x_ref, 0.1)
+    return knn_detector
+
+
+def test_unfitted_knn_single_score():
+    knn_detector = KNN(k=10)
+    x = np.array([[0, 10], [0.1, 0]])
+
+    # test predict raises exception when not fitted
+    with pytest.raises(NotFittedError) as err:
+        _ = knn_detector.predict(x)
+    assert str(err.value) == 'KNNTorch has not been fit!'
```

Comment:
Just noticed that the error message refers to the backend `KNNTorch` implementation whilst the user is calling the frontend `KNN` class, this could be confusing? Should the error message read `KNN` instead? Not sure what the implications are for refactoring where the error gets generated and then propagated...

Reply:
Perhaps the backend should generate its backend-specific error which then might be caught by the frontend which raises its own error from the backend one (so there is a full chain of exceptions)? @ascillitoe

Reply:
Yea I think so too, something like (haven't tested):
```python
# in KNN.fit()
try:
    # some calls to backend
    ...
except NotFittedError as e:
    raise NotFittedError(f'{self.__class__.__name__} has not been fit!') from e
```

`NotFittedError` could also have a parametrized constructor so that the message doesn't have to be repeated in every place, i.e. just pass an argument `class_name` and the error constructs the appropriate message.

Would be good to check and see what the stack trace looks like with this modification.